apps/aefate/src/aefa_fate.erl

```diff
@@ -0,0 +1,475 @@
+-module(aefa_fate).
+-export([run/2]).
+
+-export([get_trace/1]).
+
+-include("aefa_data.hrl").
+
+-ifdef(TEST).
+-define(trace(I,S), S#{trace => [{I, erlang:process_info(self(), reductions)} |get_trace(S)]}).
+-else.
+-define(trace(I,S), S).
+-endif.
+
+run(What, Chain) ->
+    EngineState = setup_engine(What, Chain),
+    execute(EngineState).
+
+execute(EngineState) ->
+    Instructions = current_bb(EngineState),
+    loop(Instructions, EngineState).
+
+loop(Instructions, EngineState) ->
+    case step(Instructions, EngineState) of
+        {stop, FinalState} ->
+            FinalState;
+        {jump, BB, NewState} ->
+            {NewInstructions, State2} = jump(BB, NewState),
+            loop(NewInstructions, State2)
+    end.
+
+step([], EngineState) ->
```

Comment:
Do we really want fall-through? I could imagine always having a jump instruction at the end of BBs.